### PR TITLE
fixed GitLab Discussions API request params

### PR DESCRIPTION
**Pull request by oboenikui**
fixed https://github.com/danger/danger/issues/1111

Fixed GitLab Discussions POST API params.

Although not in the GitLab API documentation, `old_line` and` old_path` seem to be required parameters when `new_path` and `new_line` were used.

### CODE REVIEW

**Comment by pbendersky** [comment]
I was sure those were optional...

What happens if a file is renamed or moved? Is `old_path` the same as `new_path`?

**Comment by oboenikui** [comment]
Hmm... That's right... Some file needs `old_line`, but other doesn't.

And comment field of my on-premises GitLab was broken when trying variously... 😂

I'll try to check GitLab CE implementations...

**Comment by oboenikui** [comment]
I checked some cases.

## TL;DR

This PR was not enough to consider. 

## Details

I'll write the result below.

- `old_line` should be empty if the noted line is included as MR differences (This case is probably normal).
- `old_line` should be filled if the file is in the MR difference but the line is unchanged.
- There may be no way to note on files that do not have differences.
- You should not be able to note on files that have just been renamed and have no differences, but noting without specifying `old_line` and` old_path` will break the discussion field. ([See this issue](https://gitlab.com/gitlab-org/gitlab-ce/issues/60262))
- In other cases, omitting `old_line` may break the discussion filed.
- If you omit `old_path`, comments will not appear in the "Changes" column. (Is this expected behavior?)

I'll consider how to fix based on these specs.

**Comment by orta** [comment]
good luck! Sorry it's tricky

**Comment by RDIL** [comment]
I fixed the merge conflict

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,10 @@
 
 ## master
 
+## Release number TBD
```
Comment (orta) [inline CHANGELOG.md]: This h2 is superfluous, it's in the master section - see above

> // Please add your own contribution below inside the "master" section, no need to
> // set a version number, that happens during a deploy.

**Comment by oboenikui** [comment]
I could possibly fix.

Since I may not have grasped the specification such as sticky correctly, could you code review the correction part of spec?

(Also, please correct my English if it is wrong.)

**Comment by pbendersky** [comment]
Looks good to me. I only left one small comment I would address.

**Comment by reshadf** [comment]
Is this going to be merged soon? :)

**Comment by oboenikui** [comment]
> Is this going to be merged soon? :)

Sorry, I don't know... If you want to test this now, add the line below to your Gemfile.

```ruby
gem 'danger', :github => "oboenikui/danger", :branch => "fix-gitlab-mr-comment"
```

**Comment by reshadf** [comment]
> > Is this going to be merged soon? :)
> 
> Sorry, I don't know... If you want to test this now, add the line below to your Gemfile.
> 
> ```ruby
> gem 'danger', :github => "oboenikui/danger", :branch => "fix-gitlab-mr-comment"
> ```

we use `gem 'danger-gitlab'` is it safe to use the danger version instead?

**Comment by oboenikui** [comment]
Since the dependencies of danger-gitlab are very simple, I think that there is no big problem for testing. (except if there is something wrong with the master branch)

https://github.com/danger/danger-gitlab-gem

**Comment by reshadf** [comment]
> Since the dependencies of danger-gitlab are very simple, I think that there is no big problem for testing. (except if there is something wrong with the master branch)
> 
> https://github.com/danger/danger-gitlab-gem

I need to use danger-gitlab.. can't use it without.

**Comment by pbendersky** [comment]
I'm ok to merge it. @orta?

**Comment by reshadf** [comment]
I tested and its still not working. I still get this output

```
BUILD SUCCESSFUL in 1m 26s
28 actionable tasks: 20 executed, 8 from cache
bundler: failed to load command: danger (/usr/share/rvm/gems/ruby-2.6.2/bin/danger)
Gitlab::Error::NotFound: Server responded with code 404, message: 404 API Version Not Found. Request URI: https://gitlab.intern.proactive.nl/api/v4http://gitlab.intern.proactive.nl/api/v4/projects/26/merge_requests/434/discussions?id=26¬eable_id=434&page=2&per_page=20
  /usr/share/rvm/gems/ruby-2.6.2/gems/gitlab-4.10.0/lib/gitlab/request.rb:67:in `validate'
  /usr/share/rvm/gems/ruby-2.6.2/gems/gitlab-4.10.0/lib/gitlab/request.rb:46:in `block (2 levels) in <class:Request>'
  /usr/share/rvm/gems/ruby-2.6.2/gems/gitlab-4.10.0/lib/gitlab/paginated_response.rb:92:in `next_page'
  /usr/share/rvm/gems/ruby-2.6.2/gems/gitlab-4.10.0/lib/gitlab/paginated_response.rb:40:in `each_page'
  /usr/share/rvm/gems/ruby-2.6.2/gems/gitlab-4.10.0/lib/gitlab/paginated_response.rb:47:in `auto_paginate'
  /usr/share/rvm/gems/ruby-2.6.2/bundler/gems/danger-55a5431104e8/lib/danger/request_sources/gitlab.rb:78:in `mr_comments'
  /usr/share/rvm/gems/ruby-2.6.2/bundler/gems/danger-55a5431104e8/lib/danger/request_sources/gitlab.rb:165:in `update_pull_request_with_inline_comments!'
  /usr/share/rvm/gems/ruby-2.6.2/bundler/gems/danger-55a5431104e8/lib/danger/request_sources/gitlab.rb:158:in `update_pull_request!'
  /usr/share/rvm/gems/ruby-2.6.2/bundler/gems/danger-55a5431104e8/lib/danger/danger_core/dangerfile.rb:247:in `post_results'
  /usr/share/rvm/gems/ruby-2.6.2/bundler/gems/danger-55a5431104e8/lib/danger/danger_core/dangerfile.rb:278:in `run'
  /usr/share/rvm/gems/ruby-2.6.2/bundler/gems/danger-55a5431104e8/lib/danger/danger_core/executor.rb:29:in `run'
  /usr/share/rvm/gems/ruby-2.6.2/bundler/gems/danger-55a5431104e8/lib/danger/commands/runner.rb:72:in `run'
  /usr/share/rvm/gems/ruby-2.6.2/gems/claide-1.0.2/lib/claide/command.rb:334:in `run'
  /usr/share/rvm/gems/ruby-2.6.2/bundler/gems/danger-55a5431104e8/bin/danger:5:in `<top (required)>'
  /usr/share/rvm/gems/ruby-2.6.2/bin/danger:23:in `load'
  /usr/share/rvm/gems/ruby-2.6.2/bin/danger:23:in `<top (required)>'
ERROR: Job failed: exit status 1
```

**Comment by orta** [comment]
@pbendersky - I trust you, merge when you want and I'll make a release when I notice 👍

**Comment by pbendersky** [comment]
@reshadf this looks different to what we are addressing in this PR... What are your GitLab environment variables?

**Comment by pbendersky** [comment]
> @pbendersky - I trust you, merge when you want and I'll make a release when I notice 👍

How do you feel about the `inline` attrbute in the `Comment` class? It makes sense in this context, but that class is used for all backends.

**Comment by orta** [comment]
I'm alright with it 👍

**Comment by reshadf** [comment]
> @reshadf this looks different to what we are addressing in this PR... What are your GitLab environment variables?

please see #1111 I followed the setup guide on danger.systems. I used the environment variables defined there

DANGER_GITLAB_HOST=git.domain.com
DANGER_GITLAB_API_BASE_URL=https://git.domain.com/api/v4
DANGER_GITLAB_API_TOKEN

**Comment by pbendersky** [comment]
I've tried different scenarios and seems to be working for me. I'll merge it, as it's better than what we currently have. If we find more errors, we can always correct them later.
